data/libs/SpaceStation.lua

```diff
@@ -275,22 +327,153 @@ end
 --
 -- Status:
 --
---   experimental
+--   stable
 --
 function SpaceStation:AddEquipmentStock (e, stock)
 	assert(self:exists())
+	assert(equipmentStock[self])
+
 	equipmentStock[self][e] = (equipmentStock[self][e] or 0) + stock
+end
+
+-- ============================================================================
+
+--
+-- Method: GetCommodityPrice
+--
+-- Get the price of a commodity item traded at this station
+--
+-- > price = station:GetCommodityPrice(itemType)
+--
+-- Parameters:
+--
+--   itemType - the <CommodityType> of the commodity item in question
+--
+-- Returns:
+--
+--   price - the price of the commodity item
+--
+-- Availability:
+--
+--   June 2022
+--
+-- Status:
+--
+--   stable
+--
+---@param itemType CommodityType
+---@return number price
+function SpaceStation:GetCommodityPrice(itemType)
+	assert(self:exists())
+
+	local price = commodityPrice[self] and commodityPrice[self][itemType.name]
+	if price then return price end
+
+	local mul = ((100 + Game.system:GetCommodityBasePriceAlterations(itemType.name)) / 100)
+	return itemType.price * mul
+end
+
+--
+-- Method: SetCommodityPrice
+--
+-- Set the price of a commodity item traded at this station
+--
+-- > station:SetCommodityPrice(itemType, price)
+--
+-- Parameters:
+--
+--   itemType - the <CommodityType> of the commodity item in question
+--
+--   price - the new price of the commodity item
+--
+-- Availability:
+--
+--   June 2022
+--
+-- Status:
+--
+--   stable
+--
+---@param itemType CommodityType
+---@param price number
+function SpaceStation:SetCommodityPrice(itemType, price)
+	assert(self:exists())
+
+	if not commodityPrice[self] then commodityPrice[self] = {} end
+	commodityPrice[self][itemType.name] = price
+end
+
+--
+-- Method: GetCommodityStock
+--
+-- Get the quantity of a cargo item this station has available for trade
+--
+-- > stock = station:GetCommodityStock(itemType)
+--
+-- Parameters:
+--
+--   itemType - the <CommodityType> of the commodity item in question
+--
+-- Returns:
+--
+--   stock - the amount available for trade
+--
+-- Availability:
+--
+--   201308
+--
+-- Status:
+--
+--   stable
+--
+---@param itemType CommodityType
+---@return integer stock
+function SpaceStation:GetCommodityStock(itemType)
+	assert(self:exists())
+
+	return (commodityStock[self] and commodityStock[self][itemType.name]) or 0
+end
+
+--
+-- Method: AddCommodityStock
+--
+-- Modify the quantity of a cargo item this station has available for trade
+--
+-- > station:AddCommodityStock(itemType, amount)
+--
+-- Parameters:
+--
+--   itemType - a <CommodityType> cargo item
+--
+--   amount - the amount of the item to add (or subtract) from the station stock
+--
+-- Availability:
+--
+--   June 2022
+--
+-- Status:
+--
+--   stable
+--
+---@param itemType CommodityType
+---@param amount integer
+function SpaceStation:AddCommodityStock(itemType, amount)
```

Comment:
hmm, all underlying data structures are indexed just by name and you use `itemType.name` 7 times in this function. Interestingly, all `CommodityTypes` are collected in a `Commodities` table indexed by names. Therefore, often, a `CommodityType` is extracted from `Commodities` by a given name, and passed to `AddCommodityStock`, which again extracts the same name from it! It doesn't seem scary, but it's kind of weird, imho.

Reply:
This is an intentional API design intended to future proof the system and reduce the number of changes to callsites. At the moment, only the ID of the commodity (the `name` field) is needed to index the underlying datastructure for commodity stocks, but future changes or improvements to the stocking algorithm may require additional introspection into the commodity type.

I've defined an (undocumented) general convention that commodities are used in public APIs by passing around a transient CommodityType instance to API functions, which internally index and serialize commodity information based on the commodity's unique ID field (`name`). This allowed existing code to be ported as easily as possible, maintains as much room for future expansion as possible, and avoids any possibility of serialization errors arising from depending on table identities remaining the same across saved games.

As an aside, even with this API design I've made the best effort possible to ensure existing code that does depend on table identities will remain functional, as I don't have the time or energy to rewrite the affected parts of multiple mission modules.

This also keeps the API design "mentally homogeneous" with future API needs for for managing equipment-as-cargo, where unique instances of equipment items will need to be added and managed in cargo holds. You'd call `cargoMgr:AddEquipment(EquipTypeInstance)`, just as you call `cargoMgr:AddCommodity(CommodityTypeInstance)` now.

Reply:
@Web-eWorks is `SpaceStation.GetStationTargetStock` part of the public API?

Reply:
It is not intended to be public (nothing uses it outside of SpaceStation), but it is intended as a low-level function operating on the "economic sim" rather than dealing with high-level Commodity objects. It currently looks like a public API function however, though I'll consider it to be the exception that proves the rule - the function is only concerned with the economy backend, hence why it takes the name of a specific Commodity coming from LuaEconomy.